nixos/modules/services/networking/netbird.nix

```diff
@@ -1,171 +1,513 @@
+    toShellVars
+    versionAtLeast
     versionOlder
     ;
 
   inherit (lib.types)
+    anything
     attrsOf
+    bool
+    enum
+    package
     port
     str
     submodule
     ;
 
-  kernel = config.boot.kernelPackages;
+  inherit (config.boot) kernelPackages;
+  inherit (config.boot.kernelPackages) kernel;
 
   cfg = config.services.netbird;
+
+  toClientList = fn: map fn (attrValues cfg.clients);
+  toClientAttrs = fn: mapAttrs' (_: fn) cfg.clients;
+
+  hardenedClients = filterAttrs (_: client: client.hardened) cfg.clients;
+  toHardenedClientList = fn: map fn (attrValues hardenedClients);
+  toHardenedClientAttrs = fn: mapAttrs' (_: fn) hardenedClients;
+
+  nixosConfig = config;
 in
 {
   meta.maintainers = with maintainers; [
     misuzu
     thubrecht
+    nazarewk
   ];
   meta.doc = ./netbird.md;
 
+  imports = [
+    (mkRemovedOptionModule [ "services" "netbird" "tunnels" ]
+      "The option `services.netbird.tunnels` has been renamed to `services.netbird.clients`")
+  ];
+
   options.services.netbird = {
-    enable = mkEnableOption "Netbird daemon";
+    enable = mkOption {
+      type = bool;
+      default = false;
+      description = ''
+        Enables backwards compatible Netbird client service.
+
+        This is strictly equivalent to:
+
+        ```nix
+        services.netbird.clients.wt0 = {
+          port = 51820;
+          name = "netbird";
+          interface = "wt0";
+          hardened = false;
+        };
+        ```
+      '';
+    };
     package = mkPackageOption pkgs "netbird" { };
 
-    tunnels = mkOption {
+    ui.enable = mkOption {
+      type = bool;
+      default = config.services.xserver.displayManager.sessionPackages != [ ];
+      defaultText = literalExpression ''config.services.xserver.displayManager.sessionPackages != [ ]'';
+      description = ''
+        Controls presence `netbird-ui` wrappers, defaults to presence of graphical sessions.
+      '';
+    };
+    ui.package = mkPackageOption pkgs "netbird-ui" { };
+
+    clients = mkOption {
       type = attrsOf (
         submodule (
           { name, config, ... }:
           {
             options = {
               port = mkOption {
                 type = port;
-                default = 51820;
+                example = literalExpression "51820";
+                description = ''
+                  Port the Netbird client listens on.
+                '';
+              };
+
+              name = mkOption {
+                type = str;
+                default = "netbird-${name}";
+                description = ''
+                  Primary name for use in:
+                  - systemd service name,
+                  - hardened user name and group,
+                  - [systemd `*Directory=`](https://www.freedesktop.org/software/systemd/man/latest/systemd.exec.html#RuntimeDirectory=) names,
+                  - desktop application identification,
+                '';
+              };
+
+              interface = mkOption {
```

Comment:
Interfaces names have a maximum length of 15 characters on linux. Since the default `netbird-${name}` is already quite long and only leaves 7 characters for the $name, can we add an assertion that this limit isn't exceeded? Maybe it would also be better to choose a shorter default prefix.

Reply:
wasn't aware of that 15 character limit, thanks for pointing it out